File: python/examples/epoch_anomalies.py
"""Epoch and anomaly conversion.

Units: MJD2000 days and radians.
Expected: arrival is 180 days later; anomaly round trip is 0.7 radians.
Runtime: constant work, normally below 1 ms with a release wheel.
Features: default wheel; no external data or runtime.
"""

import pykep_rust as pk


departure = pk.Epoch.from_iso("2030-01-15T12:30:00")
arrival = departure.add_days(180.0)
eccentric = pk.mean_to_eccentric_anomaly(0.7, 0.8)
round_trip = pk.eccentric_to_mean_anomaly(eccentric, 0.8)
assert arrival.seconds_since(departure) == 180.0 * pk.DAY_TO_SECONDS
assert abs(round_trip - 0.7) < 1e-14
print(f"departure: {departure}; arrival: {arrival}")
print(f"elliptic round trip [rad]: {round_trip:.16f}")

File: examples/src/bin/epoch-anomalies.rs
//! Convert an epoch and round-trip elliptic and hyperbolic anomalies.
//!
//! Units: epochs use MJD2000 days and anomalies are radians.
//! Expected: arrival 180 days later and round trips equal to 0.7 and 4.0.
//! Runtime: constant work, normally below 1 ms in a release build.
//! Features: default `pykep-core`; no external data or runtime.

use pykep_core::Result;
use pykep_core::astro::anomalies::{
    eccentric_to_mean_anomaly, hyperbolic_anomaly_to_mean, hyperbolic_mean_to_anomaly,
    mean_to_eccentric_anomaly,
};
use pykep_core::time::epoch::Epoch;

fn main() -> Result<()> {
    let departure = Epoch::from_iso("2030-01-15T12:30:00")?;
    let arrival = departure.checked_add_days(180.0)?;
    println!("departure: {departure}");
    println!("arrival:   {arrival}");

    let eccentric = mean_to_eccentric_anomaly(0.7, 0.8)?;
    println!(
        "elliptic round trip: {}",
        eccentric_to_mean_anomaly(eccentric, 0.8)?
    );

    let hyperbolic = hyperbolic_mean_to_anomaly(4.0, 1.5)?;
    println!(
        "hyperbolic round trip: {}",
        hyperbolic_anomaly_to_mean(hyperbolic, 1.5)?
    );
    Ok(())
}
